Implement multibackend abstraction

MortimerGoro opened the pull request:
- `com.igalia.wolvic.browser.api.*` contains the multibackend abstraction
- `com.igalia.wolvic.browser.api.impl.*` contains the implementation for Gecko
- All references to gecko have been removed from all the clasess and changed to use the abstraction directly

svillar commented on app/src/test/java/com/igalia/wolvic/ExampleUnitTest.java:
```diff
@@ -1,8 +1,8 @@
 package com.igalia.wolvic;
 
-import org.junit.Test;
+import static org.junit.Assert.assertEquals;
```
Nice, but unrelated I guess

MortimerGoro replied:
Fixed

svillar commented on app/src/common/shared/com/igalia/wolvic/browser/api/WRuntimeSettings.java:
```diff
@@ -0,0 +1,666 @@
+            return mSettings;
+        }
+        /**
+         * Set the custom Gecko process arguments.
```
Nit: replace Gecko in the docs in this method and the following ones by something like Web Engine for example

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+        if (mPermissionDelegate == null) {
+            mSession.setPermissionDelegate(null);
+        } else {
+            mSession.setPermissionDelegate(new PermissionDelegateImpl(mPermissionDelegate, this));
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+        if (mContentDelegate == null) {
+            mSession.setContentDelegate(null);
+        } else {
+            mSession.setContentDelegate(new ContentDelegateImpl(mContentDelegate, this));
```
What about using the ternary operator to avoid the if-else block and do it in one line?

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+        if (mProgressDelegate == null) {
+            mSession.setProgressDelegate(null);
+        } else {
+            mSession.setProgressDelegate(new ProgressDelegateImpl(mProgressDelegate, this));
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+            mSession.setScrollDelegate(null);
+        } else {
+            mSession.setScrollDelegate(new ScrollDelegateImpl(mScrollDelegate, this));
+        }
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+            mSession.setHistoryDelegate(null);
+        } else {
+            mSession.setHistoryDelegate(new HistoryDelegateImpl(mHistoryDelegate, this));
+        }
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+            mSession.setContentBlockingDelegate(null);
+        } else {
+            mSession.setContentBlockingDelegate(new ContentBlockingDelegateImpl(mContentBlockingDelegate, this));
+        }
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+            mSession.setPromptDelegate(null);
+        } else {
+            mSession.setPromptDelegate(new PromptDelegateImpl(mPromptDelegate, this));
+        }
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+            mSession.setMediaSessionDelegate(null);
+        } else {
+            mSession.setMediaSessionDelegate(new MediaSessionDelegateImpl(this, delegate));
+        }
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+            mSession.setSelectionActionDelegate(null);
+        } else {
+            mSession.setSelectionActionDelegate(new SelectionActionDelegateImpl(mSelectionActionDelegate, this));
+        }
```
Ditto ternary operator

MortimerGoro replied:
done

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -0,0 +1,391 @@
+
+    private int toGeckoFlags(@LoadFlags int flags) {
+        int result = 0;
+        if ((flags & WSession.LOAD_FLAGS_NONE) != 0) {
```
Do we really need the `!= 0` ? Same for the other if conditions

MortimerGoro replied:
Java doesn't convert to boolean

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:
```diff
@@ -0,0 +1,51 @@
+    @NonNull
+    @Override
+    public WResult<Bitmap> capturePixelsWithAspectPreservingSize(int width) {
+        return new ResultImpl<>(mDisplay.screenshot().aspectPreservingSize(500).capture());
```
Why 500 instead of the passed int width?

MortimerGoro replied:
Good catch!

MortimerGoro replied:
fixed

svillar commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/PromptDelegateImpl.java:
```diff
@@ -0,0 +1,844 @@
+            @Override
+            public AuthOptions authOptions() {
+                int flags = 0;
+                if ((mGeckoPrompt.authOptions.flags & AuthPrompt.AuthOptions.Flags.CROSS_ORIGIN_SUB_RESOURCE) != 0) {
```
Do we need the `!= 0` comparison? Same for the other if conditions.

MortimerGoro replied:
yes, Java doesn't convert int to boolean

svillar commented on app/src/common/shared/com/igalia/wolvic/browser/engine/Session.java:
```diff
@@ -549,30 +545,20 @@ private void restore() {
-                .suspendMediaWhenInactive(aSettings.isSuspendMediaWhenInactiveEnabled())
-                .build();
+    private WSession createISession(@NonNull SessionSettings aSettings) {
+        WSessionSettings values = WSessionSettings.create(aSettings.isPrivateBrowsingEnabled());
```
Nit: values->settings

MortimerGoro replied:
fixed

svillar commented on app/src/common/shared/com/igalia/wolvic/browser/engine/Session.java:
```diff
@@ -1484,149 +1434,115 @@ public void onContentLoaded(@NonNull GeckoSession geckoSession, @NonNull Content
-                return;
-            }
-        }
-    }*/
```
Where are these two removed now?

MortimerGoro replied:
It's a leftover from an old API. Before using MesiaSession Gecko exposed an API called MediaElement

svillar commented on app/src/common/shared/com/igalia/wolvic/browser/api/WAutocomplete.java:
```diff
@@ -0,0 +1,1301 @@
+import java.lang.annotation.RetentionPolicy;
+
+/**
+ * The Autocomplete API provides a way to leverage Gecko's input form handling for autocompletion.
```
Nit: Gecko->Browser Engine

MortimerGoro replied:
fixed

svillar reviewed (changes requested):
Amazing patch, superb documentation, pretty difficult to review :)

As a general comment, I'd very much prefer to split this at least in two patches, one of them with an initial cleanup (mostly imports) and the other one with the refactoring. The refactoring is huge by itself so the smaller the better (Ideally there would be a third one removing the Servo code but if it's too much work I'm fine including that work as part of the cleanup and/or refactoring).

There are many comments but some of them are minor and repeated over different files. There are nits here and there also, feel free to skip them if they involve a lot of work. However there are legitimate comments as well that I believe are actual bugs, please review them carefully.

Looking forward to integrating this in the tree.

MortimerGoro commented:
Thanks for the detailed review, I created the cleanup PR here https://github.com/Igalia/wolvic/pull/59 and will rebase this branch when that's merged

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/ui/widgets/SuggestionsWidget.java:
```diff
@@ -25,13 +25,14 @@
 import com.igalia.wolvic.utils.ViewUtils;
 
+import org.mozilla.geckoview.GeckoSession;
+
```
good catch, done

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/ui/widgets/dialogs/SelectionActionWidget.java:
```diff
@@ -11,7 +13,6 @@
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
-import org.mozilla.geckoview.GeckoSession;
```
good catch, done

MortimerGoro commented on app/src/test/java/com/igalia/wolvic/EnvironmentsTest.kt:
```diff
@@ -2,15 +2,15 @@ package com.igalia.wolvic
 import org.junit.runner.RunWith
-import com.igalia.wolvic.browser.SettingsStore
-import com.igalia.wolvic.utils.EnvironmentUtils
-import com.igalia.wolvic.utils.TestFileUtils
```
fixed

MortimerGoro commented on app/src/oculusvr/java/com/igalia/wolvic/PlatformActivity.java:
```diff
@@ -5,15 +5,14 @@
 import android.view.View;
 import android.view.WindowManager;
 
-import com.igalia.wolvic.browser.SettingsStore;
```
fixed

MortimerGoro commented on app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java:
```diff
@@ -25,7 +24,7 @@ public HVRSpeechRecognizer(Context context) {
 
     @Override
-    public void start(@NonNull Settings settings, @Nullable GeckoWebExecutor executor, @NonNull Callback callback) {
+    public void start(@NonNull Settings settings, @NonNull Callback callback) {
```
Fixed in a separate PR

MortimerGoro commented on app/src/hvr/java/com/igalia/wolvic/HVRLocationManager.java:
```diff
@@ -31,7 +28,7 @@ public HVRLocationManager(Context ctx) {
     private void initReflection() {
         try {
-            mMethod = GeckoAppShell.class.getDeclaredMethod("onLocationChanged", double.class, double.class, double.class, float.class, float.class, float.class, float.class, long.class);
+            mMethod = Class.forName("org.mozilla.gecko.GeckoAppShell").getDeclaredMethod("onLocationChanged", double.class, double.class, double.class, float.class, float.class, float.class, float.class, long.class);
```
I moved this out of the class

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/utils/SystemUtils.java:
```diff
@@ -10,13 +10,14 @@
 
+import org.mozilla.geckoview.CrashReporter;
+import org.mozilla.geckoview.GeckoResult;
+
```
fixed

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/crashreporting/CrashReporterService.java:
```diff
@@ -9,13 +9,14 @@
 import com.igalia.wolvic.browser.SettingsStore;
 import com.igalia.wolvic.utils.SystemUtils;
 
+import org.mozilla.geckoview.GeckoRuntime;
```
Fixed

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/crashreporting/GlobalExceptionHandler.java:
```diff
@@ -6,9 +6,10 @@
 import com.igalia.wolvic.utils.SystemUtils;
 
+import org.mozilla.gecko.CrashHandler;
+
```
fixed

MortimerGoro commented:
I have addressed the comments. In adittion to #59, I created a new https://github.com/Igalia/wolvic/pull/62 PR to remove some extra unused code before landing this.

MortimerGoro commented:
Rebased on top of https://github.com/Igalia/wolvic/pull/62 and fixed some missing GeckoSession usages. I did a search for gecko in code to confirm we don't have any more dependencies

svillar reviewed (changes requested):
We're almost there. There are just a few questions from the initial review without an answer. After that I agree we're ready to land

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/browser/api/WWebResponse.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * An {@link InputStream} containing the response body, if available. Attention: the stream must
+     * be closed whenever the app is done with it, even when the body is ignored. Otherwise the
+     * connection will not be closed until the stream is garbage collected
```
fixed

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/browser/content/TrackingProtectionStore.java:
```diff
@@ -57,10 +50,9 @@ public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, Strin
+                                   @NonNull WRuntime runtime) {
         mContext = context;
         mRuntime = runtime;
-        mContentBlockingController = mRuntime.getContentBlockingController();
```
I'll create a follow-up issue about this, it was caused by a breaking change in Gecko

MortimerGoro commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/WRuntimeSettingsImpl.java:
```diff
@@ -0,0 +1,203 @@
+package com.igalia.wolvic.browser.api.impl;
```
fixed

MortimerGoro commented on app/src/common/gecko/com/igalia/wolvic/browser/api/impl/RuntimeImpl.java:
```diff
@@ -0,0 +1,179 @@
+
+    @NonNull
+    @Override
+    public void seUptLoginPersistence(Lazy<LoginsStorage> storage) {
```
done

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/browser/api/WRuntime.java:
```diff
@@ -0,0 +1,181 @@
+
+    static WRuntime create(@NonNull Context context, @NonNull WRuntimeSettings settings) {
+        return new RuntimeImpl(context, settings);
+    }
```
I agree, I was just lazy :) I created a factory class now. In the future we can make the Factory dynamic if we want to support runtime selection of web engines. For now it's easier to statically select the engine implementation

MortimerGoro commented on app/src/common/shared/com/igalia/wolvic/browser/content/TrackingProtectionStore.java:
```diff
@@ -170,18 +162,20 @@ private void setTrackingProtectionLevel(int level) {
             settings.setCookieBehavior(policy.getCookiePolicy());
 
+            mRuntime.updateTrackingProtection(settings);
+
```
Fixed, it required some extra work to avoid having that method

MortimerGoro commented:
Thanks for the review. I pushed a new PR addressing the remaining comments
